pyro/infer/autoguide/effect.py

```diff
@@ -0,0 +1,288 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from typing import Callable, Dict, Union
+
+import torch
+from torch.distributions import biject_to, constraints
+
+import pyro.distributions as dist
+import pyro.poutine as poutine
+from pyro.distributions.distribution import Distribution
+from pyro.infer.effect_elbo import GuideMessenger
+from pyro.nn.module import PyroModule, PyroParam, pyro_method
+from pyro.poutine.runtime import get_plates
+
+from .initialization import init_to_feasible, init_to_mean
+from .utils import deep_getattr, deep_setattr, helpful_support_errors
+
+
+class AutoMessengerMeta(type(GuideMessenger), type(PyroModule)):
+    pass
+
+
+class AutoMessenger(GuideMessenger, PyroModule, metaclass=AutoMessengerMeta):
+    """
+    EXPERIMENTAL Base class for :class:`pyro.infer.effect_elbo.GuideMessenger`
+    autoguides.
+    """
+
+    @pyro_method
+    def __call__(self, *args, **kwargs):
+        # Since this guide creates parameters lazily, we need to avoid batching
+        # those parameters by a particle plate, in case the first time this
+        # guide is called is inside a particle plate. We assume all plates
+        # outside the model are particle plates.
+        self._outer_plates = get_plates()
+        try:
+            return super().__call__(*args, **kwargs)
+        finally:
+            del self._outer_plates
+
+    def call(self, *args, **kwargs):
+        """
+        Method that calls :meth:`forward` and returns parameter values of the
+        guide as a `tuple` instead of a `dict`, which is a requirement for
+        JIT tracing. Unlike :meth:`forward`, this method can be traced by
+        :func:`torch.jit.trace_module`.
+
+        .. warning::
+            This method may be removed once PyTorch JIT tracer starts accepting
+            `dict` as valid return types. See
+            `issue <https://github.com/pytorch/pytorch/issues/27743>_`.
+        """
+        result = self(*args, **kwargs)
+        return tuple(v for _, v in sorted(result.items()))
+
+    def _remove_outer_plates(self, value: torch.Tensor, event_dim: int) -> torch.Tensor:
+        """
+        Removes particle plates from initial values of parameters.
+        """
+        for f in self._outer_plates:
+            dim = f.dim - event_dim
+            if -value.dim() <= dim:
+                dim = dim + value.dim()
+                value = value[(slice(None),) * dim + (slice(1),)]
+        for dim in range(value.dim() - event_dim):
+            value = value.squeeze(0)
+        return value
+
+
+class AutoNormalMessenger(AutoMessenger):
+    """
+    EXPERIMENTAL Automatic :class:`~pyro.infer.effect_elbo.GuideMessenger` ,
+    intended for use with :class:`~pyro.infer.effect_elbo.Effect_ELBO` or
+    similar.
+
+    The mean-field posterior at any site is a transformed normal distribution.
```

Comment:
Maybe you can add a note here saying that it's directly equivalent to AutoNormal guide?

Reply:
Done.